runtime/common/src/migrations.rs

```diff
@@ -564,6 +573,105 @@ where
+        let asset_id_to_foreign_asset_storage_prefix =
+            AssetIdToForeignAsset::<Runtime>::storage_prefix();
+        let foreign_asset_to_asset_id_prefix = ForeignAssetToAssetId::<Runtime>::storage_prefix();
+
+        // Data required to migrate ForeignAsset values
+        // Read all the data into memory.
+        let asset_id_to_foreign_asset_data: Vec<_> =
+            storage_key_iter::<AssetId<Runtime>, OldLocation, Blake2_128Concat>(
+                pallet_prefix,
+                asset_id_to_foreign_asset_storage_prefix,
+            )
+            .drain()
+            .collect();
+
+        // Data required to migrate ForeignAsset keys
+        let foreign_asset_to_asset_id_data: Vec<_> =
+            storage_key_iter::<OldLocation, AssetId<Runtime>, Blake2_128Concat>(
+                pallet_prefix,
+                foreign_asset_to_asset_id_prefix,
+            )
+            .drain()
+            .collect();
+
+        let foreign_asset_to_asset_id_count = foreign_asset_to_asset_id_data.len();
+
+        let migrated_count = asset_id_to_foreign_asset_data
+            .len()
+            .saturating_add(foreign_asset_to_asset_id_data.len());
+
+        log::info!("Migrating {:?} elements", migrated_count);
+
+        // Removing older entries with old location as key
+        let mut removal_result = storage::unhashed::clear_prefix(
+            &ForeignAssetToAssetId::<Runtime>::final_prefix(),
+            None,
+            None,
+        );
+        while removal_result.maybe_cursor.is_some() {
+            removal_result = storage::unhashed::clear_prefix(
+                &ForeignAssetToAssetId::<Runtime>::final_prefix(),
+                None,
+                removal_result.maybe_cursor.as_deref(),
+            );
+        }
+
+        // Write to the new storage with removed and added fields
+        for (asset_id, old_location) in asset_id_to_foreign_asset_data {
+            if let Ok(new_location) = Runtime::ForeignAsset::try_from(old_location) {
+                AssetIdToForeignAsset::<Runtime>::insert(asset_id, new_location);
+            } else {
+                log::warn!("Location could not be converted safely to xcmV4")
+            }
+        }
+
+        for (old_location, asset_id) in foreign_asset_to_asset_id_data {
+            if let Ok(new_location) = Runtime::ForeignAsset::try_from(old_location) {
+                ForeignAssetToAssetId::<Runtime>::insert(new_location, asset_id);
+            } else {
+                log::warn!("Location could not be converted safely to xcmV4")
+            }
+        }
+
+        // One db read and one db write per element, plus the on-chain storage
+        Runtime::DbWeight::get()
+            .reads(migrated_count as u64)
+            .saturating_add(Runtime::DbWeight::get().writes(migrated_count as u64 + 3u64))
+            .saturating_add(Runtime::DbWeight::get().writes(foreign_asset_to_asset_id_count as u64))
```

Comment:
Dont you need to multiply by 2 here? (as you are writing twice per `foreign_asset_to_asset_id_count`

Reply:
The other write is calculated above in `migrated_count`